torch/serialization.py

```diff
@@ -285,6 +287,44 @@ class safe_globals(_weights_only_unpickler._safe_globals):
         >>> assert torch.serialization.get_safe_globals() == []
     """
 
+def get_unsafe_globals_in_checkpoint(f: FILE_LIKE) -> List[str]:
+    """Returns a list of strings of functions/classes in a ``torch.save`` object that are not safe for ``weights_only``.
+
+    For a given function or class ``f``, the corresponding string will be of the form
+    ``{f.__module__}.{f.__name__}``.
+
+    This function will return any GLOBALs in the checkpoint that are not in the set marked safe
+    for ``weights_only`` (either via :func:`add_safe_globals` or :class:`safe_globals` context or
+    allowlisted by ``torch`` by default).
+
+    .. note::
+        This function will statically disassemble the pickle file in the checkpoint.
+        The implication is any classes dynamically pushed onto the stack during unpickling
+        will not be included in the output.
+
+    Args:
+        f: File-like object or string containing the checkpoint object saved via ``torch.save``
+
+    Returns:
+        A list of strings corresponding to pickle GLOBALs in the checkpoint that are marked safe for ``weights_only``.
+    """
+    safe_global_strings = set(_weights_only_unpickler._get_allowed_globals().keys())
+
+    with _open_file_like(f, "rb") as opened_file:
+        if _is_zipfile(opened_file):
+            with _open_zipfile_reader(opened_file) as zip_file:
+                if _is_torchscript_zip(zip_file):
+                    raise RuntimeError(
+                        "Expected input to be a checkpoint returned by torch.save but got a torchscript checkpoint"
+                    )
+                data_file = io.BytesIO(zip_file.get_record("data.pkl"))
+                instructions = pickletools.genops(data_file)
```

Comment:
Are you sure it's safe to use it? Per https://docs.python.org/3/library/pickletools.html#pickletools.genops it looks like it might do some unsafe deserialization to retrieve the arguments...

Reply:
My understanding from the [implementation](https://github.com/python/cpython/blob/main/Lib/pickletools.py#L2268) is that it does not import any modules or call any functions serialized within the pickle file

re the decoding of the arguments, that only reads ints/strings and functions used seem to be `codecs.decode`, `struct.unpack`, `bytearray` is this potentially unsafe?

Reply:
A potential alternative would be for me to rewrite the pickletools logic, would this be preferable?

Reply:
Opened https://github.com/pytorch/pytorch/pull/139221 that does this without pickletools for consideration, that might give us a bit more confidence that nothing unsafe is being done. That said I don't think pickletools is actually doing anything blatantly unsafe.